### PR TITLE
Add notice on quantity change and update `wc/store/cart` to use thunks

### PR DESCRIPTION
**Pull request by opr**
# Description

This PR adds a notice when the quantity of an item changes unexpectedly. These are mostly edge case scenarios, however there is one legit "regular" use-case. The following changes trigger the notice.
- Shopper has a bundled item in the Cart. They update the main item, which in turn causes the bundled item's quantity to update. In this situation, there is no other clue that the bundled item's quantity updated too, so a snackbar is shown.
- An extension updates the quantity of an item in the cart during `extensionCartUpdate`. The cart is returned and the notification is shown.
- [edge case] The shopper has the Cart open in two tabs. They update item A's quantity in the first tab and then go to the second tab. In this tab, the quantity of item A has not been updated yet. In this tab, they update item B's quantity. This causes the server to return the cart (with item A's updated quantity). A notice is shown to make sure the user is aware that the quantity of an item other than the one they touched was affected.
- [edge case] The shopper adds an item that does not have a minimum/maximum quantity, then loads the Cart block. The merchant then sets a minimum/maximum quantity and then the shopper sets the quantity of an item in the cart. The item's quantity updates and a notice is shown.

# Data store changes

In addition to the above, this PR updates the actions and resolvers on the `wc/store/cart` data store to use thunks. This was required because we need to dispatch actions to the `core/notices` data store, and previously the only place we could do this was in the `wc/store/cart` reducer. It is an anti-pattern to dispatch actions inside a reducer.

Because the flow of `receiveCart` now needs to dispatch actions, we needed to turn it into a thunk (or generator). Given that the other data stores in our project use thunks, I took this route.

I had to change `apiFetchWithHeaders` in `assets/js/data/shared-controls.ts` to export both a generator and an async function. This was required because the collections data store was using `apiFetchWithHeaders` in a synchronous generator, so couldn't await promises, but `apiFetchWithHeaders` function was changed to return a promise. Instead I abstracted the logic to `doApiFetchWithHeaders` which can be called by both the async function and the generator.

# TypeScript changes

To make this work correctly with TypeScript, I created some new types:

- **`ThunkReturnType`**
  - This type informs TS of the return type of an action retrieved from `dispatch` or `useDispatch` that is a thunk.
- **`Thunk`**
  - A shorthand way of writing a function that returns a function that returns _something_
- **`ResolveSelectFromMap`**
  - This function maps the resolvers defined in `resolvers.ts` to the resolvers available on a @wordpress/data store.

I also defined some specific to the `wc/store/cart` data store:

- **`CartResolveSelectFromMap`**
  - maps the cart store's action creators to the dispatch function passed to thunks.
- **`CartDispatchFromMap`**
  - maps the cart store's resolvers to the resolveSelect function passed to thunks.

Finally, to get thunks that return values working with TS, I needed to update the `DispatchFromMap` type.

The changes observed in cb10482fdcf3602fc2a891a27d319ee0fa6ab9f2 in plain English are: if the action creator extends a function that returns a generator, then its type is a Promise that returns a generator, otherwise, if it extends `Thunk` (a function that returns a function) then its type is the type returned by the function the action creator returns. If it extends neither of these, the type is void (plain action creator).

# New functions

Please see 162577f01c6cc2f65bf34452ac8c3890f1879084 where I added `notifyIfQuantityLimitsChanged`, `notifyIfQuantityChanged`, and `notifyQuantityChanges` and see 05527e570efdb454a5fcf43f4d4bf06e4d21f404 where I added `isWithinQuantityLimits`.

- `notifyQuantityChanges` is simply a wrapper which calls both `notifyIfQuantityChanged` and `notifyIfQuantityLimitsChanged`.
- `notifyIfQuantityLimitsChanged` checks the incoming cart and compares each item to the old cart. If the quantity limits has changed, and the old quantity is outside of the limits, and that item is not currently changing quantity, we can assert that the server unexpectedly changed the quantity limits and it will be auto-updated in the client.
- `notifyQuantityChanges` checks the incoming cart and compares each item to the old cart. If the quantity has changed of an item, and that item is not currently changing quantity, we can assert that the server changed it without the user requesting it, so a notification can be shown.
- `isWithinQuantityLimits` checks to see if a `CartItem`'s current quantity is within the constraints of its `quantity_limits`.

# Tests
I updated `assets/js/data/cart/test/resolvers.ts` to test thunks now, rather than generators.

 

Fixes TBC

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [ ] E2E tests

#### Internal developer testing
<details>
  <summary>Expand to see instructions</summary>

##### Prerequisites

1. Install WooCommerce Min/Max Quantities and activate it.
2. Install WooCommerce Points and rewards and activate it.
3. Add points to your account ( WooCommerce -> Points and Rewards)
4. Create a coupon on your store.
5. Add 3 or more items to your cart. Go to the Cart block. Open the Cart block again in a **new tab**

##### Case 1 - coupons

1. Open a new tab and go to the Cart block, modify the quantity of an item on your cart in tab 2.
2. Go to tab 1, apply the coupon to your cart. Verify it is added correctly and no errors occur. Verify the quantity of the item updated in the previous step updates correctly, and that you see a notice.
3. Go to tab 2 and update the quantity of an item.
6. Go back to tab 1 and remove the coupon again. Verify it is removed correctly and no errors occur. Verify the quantity of the item updated in the previous step updates correctly, and that you see a notice.
7. Go to tab 2 and update the quantity of an item.
8. Go back to tab 1 and try to apply a coupon that does not exist. Ensure you see an error about the coupon, and verify the quantity of the item updated in the previous step updates correctly, and that you see a notice.

##### Case 2 - regular quantity updates.

1. Go to tab 2 and update the quantity of an item.
2. Go to tab 1 and modify the quantity of a different item on your cart. Verify the quantity of the item updated in the previous step updates correctly, and that you see a notice.

##### Case 5 - removing items

1. Go to tab 2 and update the quantity of an item.
2. Go to tab 1 and remove a different item from your cart. Verify it removes correctly, and that the quantity of the item updated in the previous step updates correctly, and that you see a notice.

##### Case 6 - `extensionCartUpdate`

1. Go to tab 2 and update the quantity of an item.
2. Go to tab 1 and apply your points by using the WooCommerce Points and Rewards button.
3. Verify they are added correctly, and that the quantity of the item updated in step 1 updates correctly, and that you see a notice.

##### Case 7 - Min/Max quantities.

1. Open a new tab and go to the WP Dashboard. Edit one of the items in your cart. Set a minimum quantity on the item.
2. Without reloading, go to the original cart tab. Modify the quantity of an item.
3. Check the quantity of the item with the minimum quantity has been updated correctly, and a notice shows. (Note, please ignore the red notice at the top of the page for now)
4. Do the same with a max quantity.
5. Do the same with a "Group of" quantity (so, if your item in the Cart has a quantity of 10, set the "Group of" to 6.)

###### Case 8 - Address changes

1. Go to tab 2 and update the quantity of an item.
2. Go to tab 1 and update your address using the shipping calculator.
3. Verify they address updates correctly, and that the quantity of the item updated in step 1 updates correctly, and that you see a notice.
</details>

### User facing testing

1. Add a coupon to your store.
2. Add some items to your cart and go to the Cart block.
3. Test changing the quantities of some items.
5. Test applying the coupon, it should work and you should see the discount applied.
6. Try applying a nonexistent coupon, you should see an error.
7. Ensure removing an item from your cart works correctly.
8. Continue to checkout and ensure placing the order works OK.
9. Edit a page and add the All Products Block. Ensure it loads correctly in the editor and in the front-end.
10. Try adding products to your cart from the All Products Block.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

Practically negligible performance degradation in the client while quantities are checked. Degrades linearly with number of items in the cart.

### Changelog

> Show notices to the shopper if an item in the cart's quantity is updated automatically.

### CODE REVIEW

File: assets/js/data/cart/action-types.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 export const ACTION_TYPES = {
-	RECEIVE_CART: 'RECEIVE_CART',
+	SET_CART_DATA: 'SET_CART_DATA',
```
Comment (opr) [inline assets/js/data/cart/action-types.ts:2]: the action that previously dispatched `RECEIVE_CART`  has become a thunk, and instead, eventually dispatches `SET_CART_DATA`. This is to preserve the name and notion of "receiving" a cart from the server, which works the same as before, but now performs additional side effects.

File: assets/js/previews/cart.ts [hunk 2]
```diff
@@ -51,6 +51,12 @@ export const previewCart: CartResponse = {
 			backorders_allowed: false,
 			show_backorder_badge: false,
 			sold_individually: false,
+			quantity_limits: {
```
Comment (opr) [inline assets/js/previews/cart.ts:54]: The preview cart did not have these set, so it was an incorrect representation of what would really be seen by the Cart block.

File: assets/js/data/cart/actions.ts [hunk 3]
```diff
@@ -11,33 +11,27 @@ import type {
...
  */
 import { ACTION_TYPES as types } from './action-types';
-import { STORE_KEY as CART_STORE_KEY } from './constants';
 import { apiFetchWithHeaders } from '../shared-controls';
 import type { ResponseError } from '../types';
-import { ReturnOrGeneratorYieldUnion } from '../mapped-types';
+import { DispatchFromMap, ReturnOrGeneratorYieldUnion } from '../mapped-types';
+import { CartDispatchFromMap } from './index';
+
+// `Thunks are functions that can be dispatched, similar to actions creators
```
Comment (mikejolley) [inline assets/js/data/cart/actions.ts]:
```suggestion
// Thunks are functions that can be dispatched, similar to actions creators
```

File: assets/js/data/cart/actions.ts [hunk 4]
```diff
@@ -11,33 +11,27 @@ import type {
...
 
 /**
- * Returns an action object used in updating the store with the provided items
- * retrieved from a request using the given querystring.
- *
- * This is a generic response action.
+ * An action creator that dispatches the plain action responsible for setting the cart data in the store.
  *
- * @param {CartResponse} response
+ * @param {Cart} cart the parsed cart object. (Parsed into camelCase).
```
Comment (mikejolley) [inline assets/js/data/cart/actions.ts]: Should we remove docblocks rather than maintain them, since we have TS?

File: assets/js/data/cart/actions.ts [hunk 5]
```diff
@@ -208,29 +202,29 @@ export const triggerAddedToCartEvent = ( {
...
-			cache: 'no-store',
-		} );
-		yield receiveCart( response );
-		return response;
-	} catch ( error ) {
-		yield receiveError( error );
-		// If updated cart state was returned, also update that.
-		if ( error.data?.cart ) {
-			yield receiveCart( error.data.cart );
+export const applyExtensionCartUpdate =
```
Comment (mikejolley) [inline assets/js/data/cart/actions.ts:189]: Like this. Easier to understand.

File: assets/js/data/cart/actions.ts [hunk 6]
```diff
@@ -362,32 +354,32 @@ export function* addItemToCart(
...
+
+			// If updated cart state was returned, also update that.
+			if ( error.data?.cart ) {
+				dispatch.receiveCart( error.data.cart );
+			}
 		}
-	}
-	yield itemIsPendingDelete( cartItemKey, false );
-}
+		dispatch.itemIsPendingDelete( cartItemKey, false );
```
Comment (mikejolley) [inline assets/js/data/cart/actions.ts]: You used `finally` in some earlier refactors similar to this one. Should that be done here, or should `finally` be removed? See `removeCoupon`.

File: assets/js/data/cart/actions.ts [hunk 7]
```diff
@@ -398,42 +390,40 @@ export function* removeItemFromCart(
...
+			dispatch.receiveCart( response );
+		} catch ( error ) {
+			dispatch.receiveError( error );
+
+			// If updated cart state was returned, also update that.
+			if ( error.data?.cart ) {
+				dispatch.receiveCart( error.data.cart );
+			}
+		}
+		dispatch.itemIsPendingQuantity( cartItemKey, false );
```
Comment (mikejolley) [inline assets/js/data/cart/actions.ts]: You used `finally` in some earlier refactors similar to this one. Should that be done here, or should `finally` be removed? See `removeCoupon`.

File: assets/js/data/cart/actions.ts [hunk 8]
```diff
@@ -442,38 +432,37 @@ export function* changeCartItemQuantity(
...
+			throw error;
 		}
-
-		// Re-throw the error.
-		throw error;
-	}
-	yield shippingRatesBeingSelected( false );
-	return true;
-}
+		dispatch.shippingRatesBeingSelected( false );
```
Comment (mikejolley) [inline assets/js/data/cart/actions.ts]: You used `finally` in some earlier refactors similar to this one. Should that be done here, or should `finally` be removed? See `removeCoupon`.

File: assets/js/data/cart/notify-quantity-changes.ts [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+		if ( ! quantityAboveMax && ! quantityBelowMin && ! quantityOutOfStep ) {
+			return;
+		}
+
+		if ( quantityOutOfStep ) {
+			dispatch( 'core/notices' ).createInfoNotice(
+				sprintf(
+					/* translators: %1$s is the name of the item, %2$d is the quantity of the item. %3$d is a number that the quantity must be a multiple of. */
+					__(
+						'The quantity of "%1$s" has been changed to %2$d. This product must be purchased in groups of %3$d.',
```
Comment (mikejolley) [inline assets/js/data/cart/notify-quantity-changes.ts]: 👍🏻

File: assets/js/data/shared-controls.ts [hunk 10]
```diff
@@ -114,84 +102,115 @@ const batchFetch = async ( request: APIFetchOptions ) => {
...
+ * The underlying function that actually does the fetch. This is used by both the generator (control) version of
+ * apiFetchWithHeadersControl and the async function apiFetchWithHeaders.
+ */
+const doApiFetchWithHeaders = ( options: APIFetchOptions ) =>
+	new Promise( ( resolve, reject ) => {
+		// GET Requests cannot be batched.
+		if (
+			! options.method ||
+			options.method === 'GET' ||
+			isWpVersion( '5.6', '<' )
```
Comment (mikejolley) [inline assets/js/data/shared-controls.ts]: Our min supported version exceeds this now.

File: assets/js/previews/cart.ts [hunk 11]
```diff
@@ -51,6 +51,12 @@ export const previewCart: CartResponse = {
...
+				minimum: 1,
+				maximum: 99,
```
Comment (mikejolley) [inline assets/js/previews/cart.ts:56]: Noting that there is one case in the API where maximum is `null`, rather than integer. Do you want to fix that in this PR too?

File: assets/js/data/cart/notify-quantity-changes.ts (hunk 9 above)
Comment (mikejolley) [reply assets/js/data/cart/notify-quantity-changes.ts]: I ran a few of these strings though Grammarly and it suggested switching to active voice. e.g. 

`The quantity of "test" has changed to 2. You must purchase this product in groups of 3.`

Worth using?

**Review by mikejolley — COMMENTED** [review]
This is looking good! I added some feedback.

I did come across a test case that had unexpected results and may be related.

1. Tab 1: Remove item
2. Tab 2: Edit item
3. No notice about the removed item.

And test case 2:

1. Tab 1: Remove item from cart
2. Tab2: Edit the qty of that same item in the cart.
3. No visual errors/changes.

Should we adjust this?

File: assets/js/data/cart/actions.ts (hunk 4 above)
Comment (opr) [reply assets/js/data/cart/actions.ts]: Good point, I'll undo this one and any others I did, but will create an issue for the cooldown to update the rest. Would add a lot of noise to this already large PR.

File: assets/js/data/cart/actions.ts (hunk 7 above)
Comment (opr) [reply assets/js/data/cart/actions.ts]: I've updated cases where it made sense in f237ab43bd1ebff99041d92b6aad3ca4c6b63a69. In 2e4a55254490a9919a0578274dcdbd985c820195 I added it to `removeItemFromCart` and `changeCartItemQuantity` because even if these actions error, we should still set the state to reflect that they're no longer removing or updating.

File: assets/js/data/cart/notify-quantity-changes.ts (hunk 9 above)
Comment (opr) [reply assets/js/data/cart/notify-quantity-changes.ts]: Nice, thanks for doing that. @elizaan36 could you look at the strings in this file and provide different wording if necessary?

The messages we will show are:

- The quantity of "Beanie" was changed to 2.
- The quantity of "Hoodie" was decreased to 5. This is the maximum allowed quantity.
- The quantity of "Album" was increased to 10. This is the minimum required quantity.
- The quantity of "Sunglasses" was changed to 6. You must purchase this product in groups of 3.

**Comment by elizaan36** [comment]
@opr The strings look good to me.

File: assets/js/data/cart/notify-quantity-changes.ts [hunk 12]
```diff
@@ -0,0 +1,171 @@
+/**
+ * External dependencies
+ */
+import { Cart, CartItem } from '@woocommerce/types';
+import { dispatch } from '@wordpress/data';
+import { __, sprintf } from '@wordpress/i18n';
+
+const isWithinQuantityLimits = ( cartItem: CartItem ) => {
+	return (
+		cartItem.quantity >= cartItem.quantity_limits.minimum &&
```
Comment (alexflorisca) [inline assets/js/data/cart/notify-quantity-changes.ts:17]: Will the quantity limits always be defined here? Is it possible for the `.minium and `.maxium` props to be undefined?

File: assets/js/previews/cart.ts (hunk 11 above)
Comment (opr) [reply assets/js/previews/cart.ts:56]: I changed it so that in that case, instead of maximum getting set to `null` it now gets set to `9999`

File: assets/js/data/cart/test/notify-quantity-changes.ts [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+} );
+
+/**
+ * Clones the preview cart and turns it into a `Cart`.
+ */
+const getFreshCarts = (): { oldCart: Cart; newCart: Cart } => {
+	const oldCart = mapKeys(
+		cloneDeep< CartResponse >( previewCart ),
+		( _, key ) => camelCase( key )
+	) as unknown as Cart;
```
Comment (alexflorisca) [inline assets/js/data/cart/test/notify-quantity-changes.ts:33]: Why are we using `as` twice with different values here?

File: assets/js/data/cart/test/resolvers.ts [hunk 14]
```diff
@@ -0,0 +1,54 @@
+/**
```
Comment (alexflorisca) [inline assets/js/data/cart/test/resolvers.ts:1]: Nice one for converting this to ts! 🎉

**Review by alexflorisca — COMMENTED** [review]
Nice work, glad we're changing the controls to thunks in cart to align better with other data stores. Left some inline comments and I also noticed that when adding products to cart from the all products block, the mini cart doesn't update. Probably unrelated to this PR though.

File: assets/js/data/cart/notify-quantity-changes.ts (hunk 12 above)
Comment (opr) [reply assets/js/data/cart/notify-quantity-changes.ts:17]: Based on what comes back from the API, no. There was a case where `maximum` could be `null` but I changed that in  69f9570138254b3b0a255829a067e7b31452189f

File: assets/js/data/cart/test/notify-quantity-changes.ts (hunk 13 above)
Comment (opr) [reply assets/js/data/cart/test/notify-quantity-changes.ts:33]: Without the `as unknown` it complains and says the types don't match.

```
Type 'Dictionary<number | boolean | CartResponseCouponItem[] | CartResponseShippingRate[] | CartResponseShippingAddress | ... 7 more ... | ExtensionsData>' is missing the following properties from type 'Cart': coupons, shippingRates, shippingAddress, billingAddress, and 12 more
```

but we know they do because we know a `CartResponse` with camelCase keys is equivalent to a `Cart`.

**Comment by opr** [comment]
Thanks for the review @alexflorisca

> I also noticed that when adding products to cart from the all products block, the mini cart doesn't update.

I just checked on trunk, looks like it's happening there too.

**Comment by opr** [comment]
@mikejolley and @alexflorisca 

I have made some more changes:

- Add `notifyIfRemoved` function that checks for (unexpectedly) removed items.
- Update the response code when trying to remove an invalid item to 409 ([Conflict](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409), so the client can resolve the differences)
- Updates to TS and removed unnecessary doc blocks
  - Added `CartSelectFromMap` type, which is similar to `CartResolveSelectFromMap` but does not include resolvers.
- Update JS unit tests to test for notification on unexpected item removal
- Update a PHP test to reflect new error response when invalid items are in the cart.
- Add `getItemsPendingDelete` selector in the cart data store. Necessary so we can retrieve the items slated for removal when checking if notices should be shown.

Mike please give it another go with the two cases you found while testing. Cheers
-

**Review by mikejolley — APPROVED** [review]
Excellent, test cases pass. Really good job!

File: src/StoreApi/Utilities/QuantityLimits.php [hunk 15]
```diff
@@ -24,7 +24,7 @@ public function get_cart_item_quantity_limits( $cart_item ) {
 		if ( ! $product instanceof \WC_Product ) {
 			return [
 				'minimum'     => 1,
-				'maximum'     => null,
+				'maximum'     => 9999,
```
Comment (alexflorisca) [inline src/StoreApi/Utilities/QuantityLimits.php:27]: Is this big enough? Thinking about wholesale shops for example that might sell in large quantities. Maybe we add a few more 9s?

**Review by alexflorisca — COMMENTED** [review]
Looks great, just one thing about the maximum quantity

File: src/StoreApi/Utilities/QuantityLimits.php (hunk 15 above)
Comment (opr) [reply src/StoreApi/Utilities/QuantityLimits.php:27]: In https://github.com/woocommerce/woocommerce-blocks/pull/7938/files/6c02026160322c51880d1e7618c71bf478282359#diff-1fa5b0bd80112a4242c903b56beb83d233673e7742fb0616650c782713799cf0R143 the limit is also 9999, and the line your comment refers to only happens in an edge-case, so I think it's OK :)

**Review by alexflorisca — APPROVED** [review]
Fine work sir
